R/vimpact.R

```diff
@@ -0,0 +1,435 @@
+                                focal_vaccine_delivery, countries,
+                                vaccination_years)
+
+  ## We have pop data for each age group
+  ## But coverage might be for a range of ages e.g. 1 to 100
+  ## We need total pop over this range of years
+  aggregate_pop <- coverage %>%
+    dplyr::left_join(population, by = c("country" = "country",
+                                     "year" = "year",
+                                     "gender" = "gender")) %>%
+    dplyr::filter(age_from <= age & age <= age_to) %>%
+    dplyr::group_by(coverage_set, country, year, gender, activity_type,
+                    age_from, age_to) %>%
+    dplyr::summarise(population = sum(value, na.rm = TRUE))
+
+
+  coverage %>%
+    dplyr::left_join(population, by = c("country" = "country",
+                                      "year" = "year",
+                                      "gender" = "gender")) %>%
+    dplyr::filter(age_from <= age & age <= age_to) %>%
+    dplyr::left_join(aggregate_pop, by = c("country" = "country",
+                                           "coverage_set" = "coverage_set",
+                                           "year" = "year",
+                                           "gender" = "gender",
+                                           "activity_type" = "activity_type",
+                                           "age_from" = "age_from",
+                                           "age_to" = "age_to")) %>%
+    dplyr::mutate(target =
+                    dplyr::if_else(is.na(target), population, target)) %>%
+    dplyr::mutate(fvps_source = (coverage * target * value) / population) %>%
+    dplyr::mutate(fvps =
+                    dplyr::if_else(fvps_source > value, value, fvps_source)) %>%
+    dplyr::select(country, year, vaccine, activity_type, age, fvps) %>%
+    dplyr::collect()
+}
+
+#' Get coverage data for a particular touchstone and vaccine delivery method
+#'
+#' @param con DB connection
+#' @param touchtonse Touchstone to get data for
+#' @param baseline_vaccine_delivery A list of lists containing vaccine and
+#'   activity_type (routine or campaign) describing delivery
+#' @param focal_vaccine_delivery A list of lists containing vaccine and
+#'   activity_type (routine or campaign) describing delivery
+#' @param countries Optional vector of countries to filter data
+#' @param vaccination_years Option vector of years to filter data
+#'
+#' @return Tibble of coverage data
+#' @keywords internal
+get_coverage_data <- function(con, touchstone, baseline_vaccine_delivery,
+                              focal_vaccine_delivery, countries = NULL,
+                              vaccination_years = NULL) {
+  gavi_support_level <- CONCAT <- vaccine <- activity_type <- id <- year <- NULL
+  age_from <- age_to <- name <- target <- NULL
+  delivery <-  vcapply(c(focal_vaccine_delivery, baseline_vaccine_delivery),
+                       function(x) {
```

Comment:
This always feels so boilerplatey! Wish R had lambda syntax for anon functions like this.

But maybe since we're often wanting to join these `vaccine_delivery` lists like this, worth having a helper like

```
vaccine_delivery_names <- function(vaccine_delivery) {
      vcapply(vaccine_delivery, function(x) {
          paste(x$vaccine, x$activity_type, sep = "-")
      })
}
```

Then could use here and lines 70 - 73 could become

```
focal_delivery_names <- vaccine_delivery_names(focal_vaccine_delivery)
focal_scenario <- paste0(focal_scenario_type, focal_delivery_names, sep = "-", collapse = ";")
```

and something similar for lines 74 - 77